src/business_logic/transaction/fee.rs

```diff
@@ -0,0 +1,133 @@
+    general_config: StarknetGeneralConfig,
+    tx_context: TransactionExecutionContext,
+    actual_fee: u64,
+) -> Result<CallInfo, TransactionError> {
+    if actual_fee as u64 > tx_context.max_fee {
+        return Err(TransactionError::FeeError(
+            "Actual fee exceeded max fee.".to_string(),
+        ));
+    }
+
+    let fee_token_address = general_config.starknet_os_config.fee_token_address.clone();
+
+    let calldata = [
+        general_config.block_info.sequencer_address.0.clone(),
+        Felt::from_bytes_be(&actual_fee.to_be_bytes()),
+        0.into(),
+    ]
+    .to_vec();
+
+    let entry_point_selector = Felt::from_str_radix(
+        "232670485425082704932579856502088130646006032362877466777181098476241604910",
+        10,
+    )
+    .unwrap();
+
+    let fee_transfer_call = ExecutionEntryPoint::new(
+        fee_token_address,
+        calldata,
+        entry_point_selector,
+        tx_context.account_contract_address.clone(),
+        EntryPointType::External,
+        None,
+        None,
+    );
+
+    let mut resources_manager = ExecutionResourcesManager::default();
+    fee_transfer_call
+        .execute(state, &general_config, &mut resources_manager, &tx_context)
+        .map_err(|_| TransactionError::FeeError("Fee transfer failure".to_string()))
+}
+
+// ----------------------------------------------------------------------------------------
+/// Calculates the fee of a transaction given its execution resources.
+/// We add the l1_gas_usage (which may include, for example, the direct cost of L2-to-L1
+/// messages) to the gas consumed by Cairo resource and multiply by the L1 gas price.
+// ----------------------------------------------------------------------------------------
+
+pub(crate) fn calculate_tx_fee(
+    resources: HashMap<String, Felt>,
+    gas_price: u64,
+    general_config: &StarknetGeneralConfig,
+) -> Result<u64, TransactionError> {
+    let gas_usage = resources
+        .get(&"l1_gas_usage".to_string())
+        .ok_or_else(|| TransactionError::FeeError("Invalid fee value".to_string()))?
+        .to_owned();
+
+    let l1_gas_by_cairo_usage = calculate_l1_gas_by_cairo_usage(general_config, resources)?;
+    let total_l1_gas_usage = gas_usage.to_f64().unwrap() + l1_gas_by_cairo_usage;
+
+    Ok(total_l1_gas_usage.ceil() as u64 * gas_price)
+}
+// ----------------------------------------------------------------------------------------
+/// Calculates the L1 gas consumed when submitting the underlying Cairo program to SHARP.
+/// I.e., returns the heaviest Cairo resource weight (in terms of L1 gas), as the size of
+/// a proof is determined similarly - by the (normalized) largest segment.
+// ----------------------------------------------------------------------------------------
+
+pub(crate) fn calculate_l1_gas_by_cairo_usage(
+    general_config: &StarknetGeneralConfig,
+    cairo_resource_usage: HashMap<String, Felt>,
+) -> Result<f64, TransactionError> {
```

Comment:
I'm not sure if using floats in money-related projects is really a good idea (they can be... unpredictable).

Reply:
Agree with @azteca1998. We need to have something like decimal for this.